#**Thống kê mô tả dữ liệu về rựu vang đỏ :**


##**A Chuẩn bị dữ liệu**

###**1. Nguồn gốc dữ liệu**

###Dataset from kaggle: *https://www.kaggle.com/datasets/uciml/red-wine-quality-cortez-et-al-2009*

Bộ dữ liệu này là về rượi vang đỏ "Vinho Verde" của Bồ Đào Nha. Nó thường được dùng để huấn luyện cho các loại mô hình máy học nhằm phân loại chất lượng của rượi vang đỏ. Bạn có thể tìm thêm thông tin từ trang UIC của dataset này: *https://archive.ics.uci.edu/dataset/186/wine+quality*

### **2. Import thư viện và nạp dữ liệu vào notebook**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


&nbsp;&nbsp;Ở đây tôi sử dụng cách load dataset từ file csv mà tôi đã lưu trong drive của mình. Nhưng bạn cũng có thể sử dụng nhiều phương pháp khác như tải file lên từ máy (file này sẽ bị xóa sau khi bạn đóng tab này) hoặc sử dụng trực tiếp liên kết tới trang dataset của kaggle ở phía trên.

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/SGU/Machine Learning/Bai tap lab/lab5/winequality-red.csv')
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

## **B. Thông kể dữ liệu**

### **1. Tổng quan về dữ liệu**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


**Nhận xét:** <br>
Qua những thông tin cở bản của dataset ở trên, ta có thể thấy được rằng: <br>
- Dataset có 1598 mẫu và 12 thuộc tính <br>
- Tất cả các thuộc tính đều là dữ liệu kiểu số  <br>
- Không có giá trị null nào

###**2. Tổng quan về giá trị đặc trưng của các feature có trong dữ liệu**

In [ ]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

**count**: tổng số mẫu <br>
**mean**: Giá trị trung bình cộng của dữ liệu trong một cột <br>
**std**: Độ lệch chuẩn <br>
**min / max** : Giá trị nhỏ nhất/lớn nhất <br>
**25% / 50% / 75%**: Giá trị Q1 / Q2 / Q3 trong tứ phân vị của dữ liệu

- **Độ lẹch chuẩn**: https://vi.wikipedia.org/wiki/%C4%90%E1%BB%99_l%E1%BB%87ch_chu%E1%BA%A9n <br>
- **Tứ phân vị**: https://vi.wikipedia.org/wiki/T%E1%BB%A9_ph%C3%A2n_v%E1%BB%8B

###**3. Dữ liệu trùng lặp**

In [ ]:
duplicated_rows = df[df.duplicated(keep=False)]
print(duplicated_rows)
df=df.drop_duplicates()

      fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
4               7.4             0.700         0.00             1.9      0.076   
9               7.5             0.500         0.36             6.1      0.071   
11              7.5             0.500         0.36             6.1      0.071   
22              7.9             0.430         0.21             1.6      0.106   
...             ...               ...          ...             ...        ...   
1567            7.2             0.695         0.13             2.0      0.076   
1579            6.2             0.560         0.09             1.7      0.053   
1581            6.2             0.560         0.09             1.7      0.053   
1592            6.3             0.510         0.13             2.3      0.076   
1596            6.3             0.510         0.13             2.3      0.076   

      free sulfur dioxide  

**Kết quả**

In [ ]:
print( df.duplicated().sum())
print(df.shape)

0
(1359, 12)


Dữ liệu tưởng chừng hoàn hảo ban đầu hóa ra có tổng cộng lên đến 460 mẫu trùng lặp. Sau khi xử lý các mẫu trùng lặp này bằng phương thức drop_duplicates thì dữ liệu chỉ còn lại 1359 mẫu và không còn hai mẫu nào trùng nhau nữa.

###**4. Thay thế dữ liệu và thay đổi định dạng của dữ liệu**

In [ ]:
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
5      9.4        5

Trong danh sách các thuộc tính của dataset, có một số thuộc tính có tên gồm nhiều hơn một từ nên ta cần xử lý những tên này để khi xử lý thì sẽ hợp lệ với quy tắc đặt tên của python

In [ ]:
df.rename(columns = {"fixed acidity": "fixed_acidity", "volatile acidity": "volatile_acidity",
                    "citric acid": "citric_acid", "residual sugar": "residual_sugar",
                    "chlorides": "chlorides", "free sulfur dioxide": "free_sulfur_dioxide",
                    "total sulfur dioxide": "total_sulfur_dioxide"}, inplace = True)

###**5. Dữ liệu thiếu**

In [ ]:
df.isna().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Vì dữ liệu không bị khyết ở bất kỳ đầu nên bước này ta không cần thực hiện xử lý gì cả 😊